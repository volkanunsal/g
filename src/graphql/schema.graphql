type Project @model {
  id: String
  # edge: ProjectEdge
}

type Query {
  hello: String @bang
  boo: Project
}