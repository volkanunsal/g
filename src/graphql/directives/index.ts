import compose from 'compose-function'
import {
  modelDirectiveTransformer,
  modelDirectiveTypeDefs,
} from './modelDirective'
import {
  bangDirectiveTransformer,
  bangDirectiveTypeDefs,
} from './bangDirective'

export const typeDefs = [modelDirectiveTypeDefs, bangDirectiveTypeDefs]

export const transform = compose(
  modelDirectiveTransformer,
  bangDirectiveTransformer
)
