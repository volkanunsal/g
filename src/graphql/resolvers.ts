export default {
  Query: {
    hello: () => 'world',
    boo: () => 'world',
  },
}
