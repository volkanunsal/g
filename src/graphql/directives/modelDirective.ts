import { mergeSchemas } from '@graphql-tools/merge'
import { getDirective, MapperKind, mapSchema } from '@graphql-tools/utils'
import { GraphQLSchema } from 'graphql'
import pluralize from 'pluralize'

// TODO: implement resolver for connection
// TODO: @key

function getRecord(payload: {
  model: string
  obj: undefined | Record<string, any>
  args: Record<string, any>
  ctx: Record<string, any>
  auth: Record<string, any> | undefined
}): Record<string, any> | null {
  // Apply auth rules, if they exist
  console.log('auth: ', payload.auth)
  // TODO: get the record from ORM
  return { id: 1 }
}

function listRecords(payload: {
  model: string
  obj: undefined | Record<string, any>
  args: Record<string, any>
  ctx: Record<string, any>
  auth: Record<string, any> | undefined
}): { items: Record<string, any>[]; nextToken: string | null } {
  // Apply auth rules, if they exist
  console.log('auth: ', payload.auth)
  // TODO: get the records from ORM
  const items = [{ id: 1 }]
  return { items, nextToken: null }
}

function createRecord(payload: {
  model: string
  obj: undefined | Record<string, any>
  args: Record<string, any>
  ctx: Record<string, any>
  auth: Record<string, any> | undefined
}): Record<string, any> | null {
  // Apply auth rules, if they exist
  console.log('auth: ', payload.auth)
  // TODO: get the records from ORM
  return null
}

function updateRecord(payload: {
  model: string
  obj: undefined | Record<string, any>
  args: Record<string, any>
  ctx: Record<string, any>
  auth: Record<string, any> | undefined
}): Record<string, any> | null {
  // Apply auth rules, if they exist
  console.log('auth: ', payload.auth)
  // TODO: get the records from ORM
  return null
}

function getInputType(type: string) {
  if (type.includes('Int')) return 'ModelIntInput'
  if (type.includes('ID')) return 'ModelIDInput'
  return 'ModelStringInput'
}

export function modelDirectiveTransformer(schema: GraphQLSchema) {
  // TODO: add the claims payload to the info object of the resolver
  const models = new Set()
  const auths = {}
  const fields = {}
  const connectionFields: Record<
    string,
    { fieldName: string; returnType: string }[]
  > = {}

  const newSchema = mapSchema(schema, {
    [MapperKind.OBJECT_TYPE]: (fieldConfig) => {
      const modelDirective = getDirective(schema, fieldConfig, 'model')
      if (modelDirective) {
        models.add(fieldConfig.name)

        // Save the fields in a Key-Value format.
        fields[fieldConfig.name] = Object.entries(
          fieldConfig.toConfig().fields
        ).reduce((s, kv) => ({ ...s, [kv[0]]: kv[1].type.toString() }), {})
      }
      const authDirective = getDirective(schema, fieldConfig, 'auth')
      if (authDirective) {
        auths[fieldConfig.name] = authDirective[0]
      }
      return fieldConfig
    },
    [MapperKind.FIELD]: (fieldConfig, fieldName, typeName) => {
      const connectionDirective = getDirective(
        schema,
        fieldConfig,
        'connection'
      )
      if (connectionDirective) {
        const type = fieldConfig.astNode.type
        let returnType = ''
        if (type && 'type' in type && 'name' in type.type) {
          returnType = type.type.name.value
        } else if ('name' in type) {
          returnType = type.name.value
        }

        if (returnType) {
          const connFieldsForModel = connectionFields[typeName] || []
          connFieldsForModel.push({ fieldName, returnType })
          connectionFields[typeName] = connFieldsForModel
          return null
        }
      }
      return fieldConfig
    },
  })

  // TODO: define new nodes for the missing types for Subscription
  // - onCreateXXX
  // - onUpdateXXX
  // - onDeleteXXX

  const modelArr = Array.from(models) as string[]

  const t1 = `
    ${modelArr.map((model) => {
      // Extract the fields of the model type from newSchema
      const modelFields = Object.entries(fields[model] || {})
      const fieldTypesString = modelFields.map(
        (item) => item[0] + `:` + getInputType(item[1] as string)
      )
      const connFieldsForModel = connectionFields[model] || []
      const connExtensions =
        connFieldsForModel.length > 0
          ? `type ${model} {
        ${connFieldsForModel.map(
          ({ fieldName, returnType }) => `${fieldName}(
          filter: Model${returnType}FilterInput,
          sortDirection: ModelSortDirection,
          limit: Int,
          nextToken: String
        ): Model${returnType}Connection`
        )}
      }`
          : ''

      return `
        input Model${model}FilterInput {
          ${fieldTypesString}
          and: [Model${model}FilterInput]
          or: [Model${model}FilterInput]
          not: Model${model}FilterInput
        }
        type Model${model}Connection {
          items: [${model}]
          nextToken: String
        }

        input Update${model}Input {
        	${fieldTypesString}
        }

        input Create${model}Input {
        	${fieldTypesString}
        }

        input Model${model}ConditionInput {
          ${fieldTypesString}
          and: [Model${model}ConditionInput]
          or: [Model${model}ConditionInput]
          not: Model${model}ConditionInput
        }

        ${connExtensions}
        
      `
    })}

    input ModelIDInput {
    	ne: ID
    	eq: ID
    	le: ID
    	lt: ID
    	ge: ID
    	gt: ID
    	contains: ID
    	notContains: ID
    	between: [ID]
    	beginsWith: ID
    	attributeExists: Boolean
    	attributeType: ModelAttributeTypes
    	size: ModelSizeInput
    }

    enum ModelSortDirection {
      ASC
      DESC
    }

    input ModelFloatInput {
      ne: Float
      eq: Float
      le: Float
      lt: Float
      ge: Float
      gt: Float
      between: [Float]
      attributeExists: Boolean
      attributeType: ModelAttributeTypes
    }

    input ModelBooleanInput {
      ne: Boolean
      eq: Boolean
      attributeExists: Boolean
      attributeType: ModelAttributeTypes
    }

    input ModelStringInput {
      ne: String
      eq: String
      le: String
      lt: String
      ge: String
      gt: String
      contains: String
      notContains: String
      between: [String]
      beginsWith: String
      attributeExists: Boolean
      attributeType: ModelAttributeTypes
      size: ModelSizeInput
    }

    input ModelSizeInput {
      ne: Int
      eq: Int
      le: Int
      lt: Int
      ge: Int
      gt: Int
      between: [Int]
    }

    enum ModelAttributeTypes {
      binary
      binarySet
      bool
      list
      map
      number
      numberSet
      string
      stringSet
      _null
    }

    type Query {
      ${modelArr.map(
        (model) => `
        get${model}(id: ID!): ${model}
        list${pluralize(
          model
        )}(filter: Model${model}FilterInput, limit: Int, nextToken: String): Model${model}Connection
      `
      )}
    }

    type Mutation {
      ${modelArr.map(
        (model) => `
          create${model}(input: Create${model}Input!, condition: Model${model}ConditionInput): ${model}
          update${model}(input: Update${model}Input!, condition: Model${model}ConditionInput): ${model}
        `
      )}
    }
  `
  const resolvers = {
    Query: modelArr.reduce(
      (s, model) => ({
        ...s,
        [`get${model}`]: (
          obj: undefined,
          args: Record<string, any>,
          ctx: Record<string, any>
        ) => getRecord({ model, obj, args, ctx, auth: auths[model] }),
        [`list${pluralize(model)}`]: (
          obj: undefined,
          args: Record<string, any>,
          ctx: Record<string, any>
        ) => listRecords({ model, obj, args, ctx, auth: auths[model] }),
      }),
      {}
    ),
    Mutation: modelArr.reduce(
      (s, model) => ({
        ...s,
        [`create${model}`]: (
          obj: undefined,
          args: Record<string, any>,
          ctx: Record<string, any>
        ) => createRecord({ model, obj, args, ctx, auth: auths[model] }),
        [`update${pluralize(model)}`]: (
          obj: undefined,
          args: Record<string, any>,
          ctx: Record<string, any>
        ) => updateRecord({ model, obj, args, ctx, auth: auths[model] }),
      }),
      {}
    ),
  }
  return mergeSchemas({ schemas: [newSchema], typeDefs: [t1], resolvers })
}

export const modelDirectiveTypeDefs = /*graphql*/ `
  directive @key(fields: [String!]!, name: String, queryField: String) on OBJECT
  directive @connection(name: String, keyName: String, fields: [String!]) on FIELD_DEFINITION


  directive @model(
    queries: ModelQueryMap
    mutations: ModelMutationMap
    subscriptions: ModelSubscriptionMap
    timestamps: TimestampConfiguration
  ) on OBJECT
  input ModelMutationMap {
    create: String
    update: String
    delete: String
  }
  input ModelQueryMap {
    get: String
    list: String
  }
  input ModelSubscriptionMap {
    onCreate: [String]
    onUpdate: [String]
    onDelete: [String]
    level: ModelSubscriptionLevel
  }
  enum ModelSubscriptionLevel {
    off
    public
    on
  }
  input TimestampConfiguration {
    createdAt: String
    updatedAt: String
  }

  # -----------------
  # When applied to a type, augments the application with
  # owner and group-based authorization rules.
  directive @auth(rules: [AuthRule!]!) on OBJECT | FIELD_DEFINITION
  input AuthRule {
    allow: AuthStrategy!
    provider: AuthProvider
    ownerField: String # defaults to "owner" when using owner auth
    identityClaim: String # defaults to "username" when using owner auth
    groupClaim: String # defaults to "cognito:groups" when using Group auth
    groups: [String]  # Required when using Static Group auth
    groupsField: String # defaults to "groups" when using Dynamic Group auth
    operations: [ModelOperation] # Required for finer control
  }
  enum AuthStrategy { owner groups private public }
  enum AuthProvider { apiKey iam oidc userPools }
  enum ModelOperation { create update delete read }
`
