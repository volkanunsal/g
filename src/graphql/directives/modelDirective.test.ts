import { makeExecutableSchema } from '@graphql-tools/schema'
import { removeLoc } from '@graphql-tools/optimize'
import { transform, typeDefs } from './modelDirective'
import { GraphQLSchema, ObjectTypeDefinitionNode } from 'graphql'

describe('modelDirective', () => {
  function getSchema(typeDefsFromFile = '', resolvers = {}) {
    return transform(
      makeExecutableSchema({
        typeDefs: typeDefs.concat(typeDefsFromFile),
        resolvers,
      })
    )
  }

  function removeLocAny<T>(ast: T): T {
    return removeLoc(ast as any) as any
  }

  describe('@model', () => {
    let schema = getSchema(`
      type Project @model {
        id: ID
        name: String
      }
    `)

    describe('root queries', () => {
      const getQueryAST = (schema: GraphQLSchema) =>
        removeLocAny(
          schema.getTypeMap().Query.astNode as ObjectTypeDefinitionNode
        )

      const ast = getQueryAST(schema)

      it('adds: listProjects', () => {
        expect(ast.fields[1]).toMatchSnapshot()
        expect(ast.fields[1]).toHaveProperty('name.value', 'listProjects')
      })

      it('adds: getProject', () => {
        expect(ast.fields[0]).toMatchSnapshot()
        expect(ast.fields[0]).toHaveProperty('name.value', 'getProject')
      })
    })

    describe('root mutations', () => {
      const getMutationAST = (schema: GraphQLSchema) =>
        removeLocAny(
          schema.getTypeMap().Mutation.astNode as ObjectTypeDefinitionNode
        )

      const ast = getMutationAST(schema)

      it('adds: createProject', () => {
        expect(ast.fields[0]).toMatchSnapshot()
        expect(ast.fields[0]).toHaveProperty('name.value', 'createProject')
      })

      it('adds: updateProject', () => {
        expect(ast.fields[1]).toMatchSnapshot()
        expect(ast.fields[1]).toHaveProperty('name.value', 'updateProject')
      })

      it('adds: deleteProject', () => {
        expect(ast.fields[2]).toMatchSnapshot()
        expect(ast.fields[2]).toHaveProperty('name.value', 'deleteProject')
      })
    })

    describe('adds: model types', () => {
      it('adds: ModelProjectFilterInput', () => {
        expect(schema.getTypeMap()).toHaveProperty(
          'ModelProjectFilterInput.name',
          'ModelProjectFilterInput'
        )
      })

      it('adds: ModelProjectConnection', () => {
        expect(schema.getTypeMap()).toHaveProperty(
          'ModelProjectConnection.name',
          'ModelProjectConnection'
        )
      })

      it('adds: ModelProjectConditionInput', () => {
        expect(schema.getTypeMap()).toHaveProperty(
          'ModelProjectConditionInput.name',
          'ModelProjectConditionInput'
        )
      })
    })
  })

  describe('@auth', () => {
    it.todo('works')
  })

  describe('@connection', () => {
    it.todo('works')
  })

  // TODO: test resolvers
})
