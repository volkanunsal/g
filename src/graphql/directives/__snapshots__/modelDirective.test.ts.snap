// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modelDirective @model root mutations adds: createProject 1`] = `
Object {
  "arguments": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "input",
      },
      "type": Object {
        "kind": "NonNullType",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "CreateProjectInput",
          },
        },
      },
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "condition",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "ModelProjectConditionInput",
        },
      },
    },
  ],
  "description": undefined,
  "directives": Array [],
  "kind": "FieldDefinition",
  "name": Object {
    "kind": "Name",
    "value": "createProject",
  },
  "type": Object {
    "kind": "NamedType",
    "name": Object {
      "kind": "Name",
      "value": "Project",
    },
  },
}
`;

exports[`modelDirective @model root mutations adds: deleteProject 1`] = `
Object {
  "arguments": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "input",
      },
      "type": Object {
        "kind": "NonNullType",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "DeleteProjectInput",
          },
        },
      },
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "condition",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "ModelProjectConditionInput",
        },
      },
    },
  ],
  "description": undefined,
  "directives": Array [],
  "kind": "FieldDefinition",
  "name": Object {
    "kind": "Name",
    "value": "deleteProject",
  },
  "type": Object {
    "kind": "NamedType",
    "name": Object {
      "kind": "Name",
      "value": "Project",
    },
  },
}
`;

exports[`modelDirective @model root mutations adds: updateProject 1`] = `
Object {
  "arguments": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "input",
      },
      "type": Object {
        "kind": "NonNullType",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "UpdateProjectInput",
          },
        },
      },
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "condition",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "ModelProjectConditionInput",
        },
      },
    },
  ],
  "description": undefined,
  "directives": Array [],
  "kind": "FieldDefinition",
  "name": Object {
    "kind": "Name",
    "value": "updateProject",
  },
  "type": Object {
    "kind": "NamedType",
    "name": Object {
      "kind": "Name",
      "value": "Project",
    },
  },
}
`;

exports[`modelDirective @model root queries adds: getProject 1`] = `
Object {
  "arguments": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "id",
      },
      "type": Object {
        "kind": "NonNullType",
        "type": Object {
          "kind": "NamedType",
          "name": Object {
            "kind": "Name",
            "value": "ID",
          },
        },
      },
    },
  ],
  "description": undefined,
  "directives": Array [],
  "kind": "FieldDefinition",
  "name": Object {
    "kind": "Name",
    "value": "getProject",
  },
  "type": Object {
    "kind": "NamedType",
    "name": Object {
      "kind": "Name",
      "value": "Project",
    },
  },
}
`;

exports[`modelDirective @model root queries adds: listProjects 1`] = `
Object {
  "arguments": Array [
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "filter",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "ModelProjectFilterInput",
        },
      },
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "limit",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "Int",
        },
      },
    },
    Object {
      "defaultValue": undefined,
      "description": undefined,
      "directives": Array [],
      "kind": "InputValueDefinition",
      "name": Object {
        "kind": "Name",
        "value": "nextToken",
      },
      "type": Object {
        "kind": "NamedType",
        "name": Object {
          "kind": "Name",
          "value": "String",
        },
      },
    },
  ],
  "description": undefined,
  "directives": Array [],
  "kind": "FieldDefinition",
  "name": Object {
    "kind": "Name",
    "value": "listProjects",
  },
  "type": Object {
    "kind": "NamedType",
    "name": Object {
      "kind": "Name",
      "value": "ModelProjectConnection",
    },
  },
}
`;
